src/openms/source/FORMAT/MzTabMFile.cpp

```diff
@@ -0,0 +1,678 @@
+                                                              n_sml_section_columns));
+
+      OPENMS_POSTCONDITION(n_sml_header_columns == n_sml_section_columns,
+                           "The number of columns of the small molecule header do not assort to the number of columns of the small molecule section row.")
+    }
+
+    size_t n_smf_header_columns = 0;
+
+    out.emplace_back("");
+    out.emplace_back(generateMzTabMSmallMoleculeFeatureHeader_(mztab_m.getMetaData(),
+                                                               mztab_m.getMSmallMoleculeFeatureOptionalColumnNames(),
+                                                               n_smf_header_columns));
+
+    size_t n_smf_section_columns = 0;
+    const MzTabMSmallMoleculeFeatureSectionRows& feature_section = mztab_m.getMSmallMoleculeFeatureSectionRows();
+    for (const auto& smf_row : feature_section)
+    {
+      out.emplace_back(generateMzTabMSmallMoleculeFeatureSectionRow_(smf_row,
+                                                                     mztab_m.getMSmallMoleculeFeatureOptionalColumnNames(),
+                                                                     n_smf_section_columns));
+
+      OPENMS_POSTCONDITION(n_smf_header_columns == n_smf_section_columns,
+                           "The number of columns of the small molecule feature header do not assort to the number of columns of the small molecule feature section row.")
+    }
+
+    size_t n_sme_header_columns = 0;
+    out.emplace_back("");
+    out.emplace_back(generateMzTabMSmallMoleculeEvidenceHeader_(mztab_m.getMetaData(),
+                                                                mztab_m.getMSmallMoleculeEvidenceOptionalColumnNames(),
+                                                                n_sme_header_columns));
+
+    size_t n_sme_section_columns = 0;
+    const MzTabMSmallMoleculeEvidenceSectionRows& evidence_section = mztab_m.getMSmallMoleculeEvidenceSectionRows();
+    for (const auto& sme_row : evidence_section)
+    {
+      out.emplace_back(generateMzTabMSmallMoleculeEvidenceSectionRow_(sme_row,
+                                                                      mztab_m.getMSmallMoleculeEvidenceOptionalColumnNames(),
+                                                                      n_sme_section_columns));
+
+      OPENMS_POSTCONDITION(n_sme_header_columns == n_sme_section_columns,
+                           "The number of columns in the small molecule evidence header do not assort to the number of columns of the small molecule evidence section row.")
+    }
+
+    TextFile tmp_out;
+    for (TextFile::ConstIterator it = out.begin(); it != out.end(); ++it)
+    {
+      tmp_out.addLine(*it);
+    }
+    tmp_out.store(filename);
+  }
+
+  void MzTabMFile::addOptionalColumnsToSectionRow_(const std::vector<String>& column_names, const std::vector<MzTabOptionalColumnEntry>& column_entries, StringList& output) const
```

Comment:
Is this duplicated between MzTab and MzTabM?

Reply:
Yes, correct, that is some minor duplication, but I wanted to keep the main implementation separate without adding another "base class", which IMHO would be overkill here. 

I probably could add MzTabMFile as a friend to MzTabFile, to allow the usage of the protected member function, but I think keeping them separate with some (minor) duplication is the cleaner option here.

Reply:
You've already made the `MzTabBase` class. Wouldn't it fit there?

Reply:
Hm, that would then be a mix of the mzTab specification data structure and the actual mztab file (reading/writing). I think I would not do that. @timosachsenberg What is your take on that?

Reply:
Alternatively, would it make sense to make the function public?